# Week 3: Python Sorting & Calculated Fields - Practice Exercises
## Nigerian E-commerce Business Scenarios

**Course**: PORA Academy Cohort 5 - Data Analytics & AI Bootcamp  
**Assignment**: Week 3 Python Practice  
**Due Date**: Before Thursday's SQL class  

---

## 📋 Instructions for Students

You are working as a data analyst for **"Olist Nigeria,"** a major e-commerce marketplace. Complete the following exercises to help the business make data-driven decisions.

### Submission Requirements:
1. Complete all exercises with proper pandas code
2. Include comments explaining your business logic
3. Use proper variable names and code formatting
4. Handle missing values appropriately
5. Test your code and verify results make business sense

### Dataset Information:
You'll be working with Olist e-commerce data containing:
- **Orders**: order_id, price, freight_value
- **Products**: product_id, category, weight, dimensions
- **Business Context**: Brazilian marketplace data adapted for Nigerian analysis

## 🔧 Setup and Data Loading

In [1]:
# Import required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime
import warnings
warnings.filterwarnings('ignore')

# Set display options
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.2f}'.format)

print("📦 Libraries imported successfully!")

📦 Libraries imported successfully!


In [4]:
np.random.seed(42)


n_orders = 200

# Generate realistic e-commerce data
order_data = {
    'order_id': [f'order_{i:06d}' for i in range(n_orders)],
    'order_item_id': np.random.randint(1, 4, n_orders),
    'product_id': [f'prod_{np.random.randint(1000, 9999):04d}' for _ in range(n_orders)],
    'seller_id': [f'seller_{np.random.randint(100, 999):03d}' for _ in range(n_orders)],
    'price': np.random.exponential(120, n_orders).round(2),
    'freight_value': np.random.exponential(18, n_orders).round(2),
    'product_category_name_english': np.random.choice([
        'health_beauty', 'computers_accessories', 'watches_gifts', 'toys',
        'electronics', 'bed_bath_table', 'sports_leisure', 'furniture_decor',
        'housewares', 'auto', 'telephony', 'garden_tools', 'food'
    ], n_orders),
    'product_weight_g': np.random.lognormal(6, 1, n_orders), # Removed .astype(int) to allow NaNs
    'product_length_cm': np.random.randint(10, 80, n_orders),
    'product_height_cm': np.random.randint(5, 40, n_orders),
    'product_width_cm': np.random.randint(8, 60, n_orders)
}

# Add some missing values (realistic scenario)
missing_weight = np.random.choice(n_orders, size=int(n_orders * 0.08), replace=False)
for idx in missing_weight:
    order_data['product_weight_g'][idx] = np.nan

df = pd.DataFrame(order_data)

print(f"📊 Dataset loaded: {df.shape[0]} rows, {df.shape[1]} columns")
print("\n🔍 First few rows:")
df.head()

📊 Dataset loaded: 200 rows, 11 columns

🔍 First few rows:


order_id  order_item_id product_id   seller_id  price  freight_value  \
0  order_000000              3  prod_4763  seller_919 167.50          24.83   
1  order_000001              1  prod_6232  seller_755 188.18           3.67   
2  order_000002              3  prod_2853  seller_368 187.06           4.23   
3  order_000003              3  prod_7585  seller_469  11.48           8.33   
4  order_000004              1  prod_2291  seller_735  81.85          11.93   

  product_category_name_english  product_weight_g  product_length_cm  \
0                    housewares            188.74                 67   
1                          auto            402.08                 35   
2                          toys            376.54                 55   
3                    housewares             30.62                 38   
4                   electronics            230.13                 13   

   product_height_cm  product_width_cm  
0                 27                26  
1                 35                43  
2                 15                46  
3                 14                17  
4                  7                32

## 📚 Exercise 1: Basic Calculated Fields (15 points)

### Business Scenario:
The marketing team needs to understand total order values, shipping costs, and Nigerian market pricing.

### Your Task:
Create calculated columns for:
1. **Total order value** (price + freight)
2. **Estimated Nigerian Naira prices** (use 1 BRL = 500 NGN)
3. **Freight as percentage** of product price (round to 2 decimals)
4. **Business category**: "High Value" if total > 200 BRL, "Standard" otherwise
5. **Platform commission** (10% of product price)

### Requirements:
- Handle any division by zero cases
- Sort results by total order value (highest first)
- Display top 15 results with relevant columns

In [7]:
# 💰 EXERCISE 1: Your solution here

# Step 1: Calculate total order value
# YOUR CODE HERE
df['total_order_value'] = df['price'] + df['freight_value']


# Step 2: Convert to Nigerian Naira
# YOUR CODE HERE
Exchange_rate = 500
df['price_ngn'] = (df['price'] * Exchange_rate).round(0)
df['freight_value_ngn'] = (df['freight_value'] * Exchange_rate).round(0)
df['total_order_value_ngn'] = (df['total_order_value'] * Exchange_rate).round(0)

# Step 3: Calculate freight percentage
# YOUR CODE HERE
# Handle division by zero by replacing 0 price_ngn with NaN before division
df['freight_percentage'] = np.where(df['price_ngn'] == 0, np.nan, (df['freight_value_ngn']/df['price_ngn']*100).round(2))

# Step 4: Create business category
# YOUR CODE HERE
# Corrected: Business category based on total_order_value > 200 BRL
df['business_category'] = np.where(df['total_order_value'] > 200, 'High Value', 'Standard Value')

# Step 5: Calculate platform commission
# YOUR CODE HERE
df['platform_commission'] = df['price'] * 0.10

# Step 6: Sort and display results
# YOUR CODE HERE
highest_value = df.sort_values('total_order_value', ascending=False)

print("💰 Exercise 1 Results:")
# Display your results here
print("\nHighest Value Orders (Top 15):")
print(highest_value.head(15)[['order_id', 'price', 'freight_value', 'total_order_value', 'price_ngn', 'freight_value_ngn', 'total_order_value_ngn', 'freight_percentage', 'business_category', 'platform_commission']])

💰 Exercise 1 Results:

Highest Value Orders (Top 15):
         order_id  price  freight_value  total_order_value  price_ngn  \
112  order_000112 980.69          30.33            1011.02  490345.00   
113  order_000113 683.38          19.70             703.08  341690.00   
56   order_000056 594.82          72.70             667.52  297410.00   
27   order_000027 512.25          16.21             528.46  256125.00   
99   order_000099 523.71           2.39             526.10  261855.00   
80   order_000080 514.06          11.81             525.87  257030.00   
22   order_000022 517.86           7.00             524.86  258930.00   
162  order_000162 469.64          14.08             483.72  234820.00   
59   order_000059 418.03          62.80             480.83  209015.00   
52   order_000052 420.30          40.87             461.17  210150.00   
79   order_000079 439.80           3.02             442.82  219900.00   
103  order_000103 420.45           1.24             421.69  210225.00 

## 📊 Exercise 2: Multi-Column Sorting (15 points)

### Business Scenario:
The operations team needs to prioritize order fulfillment based on multiple criteria.

### Your Task:
Create a fulfillment priority system:
1. **Priority 1**: Electronics/Computers over 300 BRL
2. **Priority 2**: Health/Beauty over 150 BRL
3. **Priority 3**: All other orders over 100 BRL
4. **Priority 4**: Everything else

### Requirements:
- Create a priority column with numerical values (1-4)
- Sort by priority level, then total value (highest first), then freight efficiency (lowest % first)
- Show top 25 orders with order_id, category, price, freight_value, total_value, freight_percentage, priority
- Include a priority description column

In [ ]:
# 🎯 EXERCISE 2: Your solution here

# Step 1: Create priority function or logic
def assign_priority(row):
    """Assign fulfillment priority based on category and price"""
    # YOUR CODE HERE
    pass

# Step 2: Apply priority assignment
# YOUR CODE HERE

# Step 3: Create priority descriptions
# YOUR CODE HERE

# Step 4: Sort by multiple criteria
# YOUR CODE HERE

# Step 5: Display results
print("🎯 Exercise 2 Results - Fulfillment Priority System:")
# Display your results here

## 🧠 Exercise 3: Complex Conditional Logic (20 points)

### Business Scenario:
Management wants a comprehensive product categorization system for strategic planning.

### Your Task:
Create multiple categorization systems:

1. **Price Tier Classification:**
   - Budget: < 50 BRL
   - Economy: 50-150 BRL
   - Premium: 150-500 BRL
   - Luxury: > 500 BRL

2. **Shipping Efficiency Score:**
   - Excellent: Freight < 10% of price
   - Good: Freight 10-20% of price
   - Average: Freight 20-30% of price
   - Poor: Freight > 30% of price

3. **Weight Category:**
   - Light: < 500g
   - Medium: 500-2000g
   - Heavy: 2000-5000g
   - Bulk: > 5000g
   - Unknown: Missing weight data

4. **Business Recommendation:**
   - "Promote Heavily": Premium/Luxury with Excellent/Good shipping
   - "Standard Marketing": Economy with Good shipping
   - "Review Logistics": Any product with Poor shipping
   - "Bundle Opportunity": Budget items with Average/Good shipping
   - "Standard Treatment": All others

### Requirements:
- Use appropriate pandas methods (apply, np.where, or custom functions)
- Handle missing weight values properly
- Sort by Business Recommendation priority, then Price Tier, then Shipping Efficiency
- Display 30 results with all categorization columns

In [ ]:
# 🧠 EXERCISE 3: Your solution here

# Step 1: Price tier classification
def classify_price_tier(price):
    """Classify product by price tier"""
    # YOUR CODE HERE
    pass

# Step 2: Shipping efficiency score
def classify_shipping_efficiency(freight_pct):
    """Classify shipping efficiency"""
    # YOUR CODE HERE
    pass

# Step 3: Weight category
def classify_weight_category(weight):
    """Classify product by weight"""
    # YOUR CODE HERE
    pass

# Step 4: Business recommendation
def get_business_recommendation(row):
    """Generate business recommendation based on multiple factors"""
    # YOUR CODE HERE
    pass

# Apply all categorizations
# YOUR CODE HERE

# Sort and display results
print("🧠 Exercise 3 Results - Comprehensive Product Categorization:")
# Display your results here

# Bonus: Show distribution of each category
print("\n📊 Category Distributions:")
# Show value_counts for each category

## 🔤 Exercise 4: String Manipulation & Data Formatting (20 points)

### Business Scenario:
The finance team needs professional, formatted reports for investor presentations.

### Your Task:
Create formatted display fields:

1. **Order code**: First 8 characters of order_id in uppercase + "..."
2. **Category display**: Replace underscores with spaces, title case
3. **Price displays**:
   - BRL format: "R$ X.XX"
   - NGN format: "₦ X,XXX" (no decimals)
4. **Weight display**: "XXXg" or "Weight not specified" if missing
5. **Dimensions display**: "Length×Height×Width cm"
6. **Volume calculation**: Length × Height × Width (cubic cm)
7. **Data completeness score**: Count of non-missing fields (0-4 scale)
8. **Quality flag**:
   - "Complete": All data available
   - "Missing Weight": Weight missing but dimensions available
   - "Incomplete Data": Multiple missing fields

### Requirements:
- Use appropriate string methods (.str.upper(), .str.replace(), etc.)
- Handle missing values gracefully
- Use lambda functions or apply() where appropriate
- Sort by data completeness (complete records first), then by price
- Display 20 results with all formatted fields

In [ ]:
# 🔤 EXERCISE 4: Your solution here

# Step 1: Create order code
# YOUR CODE HERE

# Step 2: Format category display
# YOUR CODE HERE

# Step 3: Create price displays
# YOUR CODE HERE

# Step 4: Weight display
# YOUR CODE HERE

# Step 5: Dimensions display and volume
# YOUR CODE HERE

# Step 6: Data completeness score
def calculate_completeness_score(row):
    """Calculate data completeness score"""
    # YOUR CODE HERE
    pass

# Step 7: Quality flag
def assign_quality_flag(row):
    """Assign quality flag based on available data"""
    # YOUR CODE HERE
    pass

# Apply all formatting
# YOUR CODE HERE

# Sort and display results
print("🔤 Exercise 4 Results - Professional Data Formatting:")
# Display your results here

## 🚀 Exercise 5: Advanced Business Analysis (30 points)

### Business Scenario:
The CEO wants a comprehensive executive dashboard combining all Week 3 concepts.

### Your Task:
Create an executive summary combining:

1. **Financial Metrics:**
   - Total order value in both BRL and NGN
   - Estimated profit (assume 25% gross margin on price)
   - Platform commission and processing fees

2. **Efficiency Metrics:**
   - Price per gram (if weight available)
   - Value density (price per cubic cm)
   - Shipping efficiency percentage

3. **Business Classifications:**
   - Market segment (Mass/Premium/Luxury)
   - Operational priority (High/Medium/Low)
   - Growth potential assessment

4. **Executive Insights:**
   - Profitability index (custom formula)
   - Strategic recommendation
   - Risk assessment

### Requirements:
- Combine multiple calculation methods from previous exercises
- Handle all edge cases (missing data, division by zero)
- Create meaningful business categories
- Sort by strategic importance for business
- Include summary statistics and insights
- Present top 25 most strategically important products

### Bonus Challenge:
- Create a simple visualization showing key business metrics
- Write a brief executive summary of findings

In [ ]:
# 🚀 EXERCISE 5: Your comprehensive solution here

# Step 1: Financial Metrics
def calculate_financial_metrics(df):
    """Calculate comprehensive financial metrics"""
    # YOUR CODE HERE
    pass

# Step 2: Efficiency Metrics
def calculate_efficiency_metrics(df):
    """Calculate efficiency and performance metrics"""
    # YOUR CODE HERE
    pass

# Step 3: Business Classifications
def assign_business_classifications(df):
    """Assign strategic business classifications"""
    # YOUR CODE HERE
    pass

# Step 4: Executive Insights
def generate_executive_insights(df):
    """Generate executive-level insights and recommendations"""
    # YOUR CODE HERE
    pass

# Apply all analysis functions
# YOUR CODE HERE

# Strategic sorting and ranking
# YOUR CODE HERE

print("🚀 Exercise 5 Results - Executive Business Dashboard:")
# Display your comprehensive results here

In [ ]:
# 📊 BONUS: Create executive dashboard visualization

# Create a multi-panel dashboard
plt.figure(figsize=(15, 10))

# YOUR VISUALIZATION CODE HERE
# Suggestions:
# - Profit margin distribution
# - Strategic recommendations pie chart
# - Price vs efficiency scatter plot
# - Category performance comparison

plt.tight_layout()
plt.show()

print("📊 Executive Dashboard Created!")

## 💭 Reflection Questions

Answer these questions in the markdown cell below:

1. **Business Insights**: What key business insights did you discover while analyzing this e-commerce data?

2. **Data Quality**: How did missing values impact your analysis, and how did you handle them?

3. **Calculated Fields**: Which calculated fields provided the most valuable business insights?

4. **Sorting Strategy**: How did multi-column sorting help prioritize business decisions?

5. **Nigerian Market**: What considerations would be important when adapting this analysis for the Nigerian e-commerce market?

6. **SQL Comparison**: How do you think these pandas operations will translate to SQL syntax tomorrow?

---

### Your Answers:

**Your reflection answers here:**

1. **Business Insights**:
   [Your answer here]

2. **Data Quality**:
   [Your answer here]

3. **Calculated Fields**:
   [Your answer here]

4. **Sorting Strategy**:
   [Your answer here]

5. **Nigerian Market**:
   [Your answer here]

6. **SQL Comparison**:
   [Your answer here]

## 📝 Submission Checklist

Before submitting, verify:

- [ ] All exercises completed with working code
- [ ] Results make business sense and are properly formatted
- [ ] Missing values handled appropriately
- [ ] Comments explain business logic clearly
- [ ] Sorting follows exercise requirements
- [ ] Code is clean and well-organized
- [ ] Reflection questions answered thoughtfully
- [ ] Visualizations (if included) are clear and informative

---

## 🎯 Assessment Rubric

**Technical Accuracy (40%)**:
- Correct pandas syntax and operations
- Proper handling of data types and missing values
- Accurate calculations and logic

**Business Application (35%)**:
- Meaningful categorization and business logic
- Appropriate handling of e-commerce scenarios
- Strategic thinking in analysis approach

**Problem Solving (15%)**:
- Creative solutions to complex requirements
- Effective use of pandas methods
- Handling edge cases and data quality issues

**Code Quality (10%)**:
- Clean, readable, well-commented code
- Proper function definitions and variable names
- Professional presentation of results

---

**Good luck! Remember: Think like a business analyst, not just a programmer. Your analysis should drive actionable business decisions! 🚀**